File: python_files/ex7.py
class Professor:
    def __init__(self,nome):
        self.nome = nome
        self.escolas = []  # Um professor pode ter várias escolas

    def adicionar_escola(self,escola):
        if escola not in self.escolas:
            self.escolas.append(escola)
            escola.adicionar_professor(self)

    def info(self):
        escolas_nomes = ', '.join([escola.nome for escola in self.escolas])
        return f"Professor {self.nome} leciona em: {escolas_nomes}"
    
class Escola:
    def __init__(self, nome):
        self.nome = nome
        self.professores = []  # Uma escola pode ter vários professores

    def adicionar_professor(self,professor):
        if professor not in self.professores:
            self.professores.append(professor)
            professor.adicionar_escola(self)

    def info(self):
        professores_nomes = ', '.join([prof.nome for prof in self.professores])
        return f"Escola {self.nome} tem os professores: {professores_nomes}"
    
# Teste
prof1 = Professor("João")
prof2 = Professor("Maria")

escola1 = Escola("Escola A")
escola2 = Escola("Escola B")

escola1.adicionar_professor(prof1)
escola2.adicionar_professor(prof1)
prof2.adicionar_escola(escola1)

print(escola1.info())
print(escola2.info())
print(prof1.info())
print(prof2.info())

File: go_files/ex7.go
package main

import "fmt"

type Professor struct {
	Nome     string
	Escolas  []*Escola
}

type Escola struct {
	Nome       string
	Professores []*Professor
}

func (p*Professor) AdicionarEscola(e *Escola) {
	for _, escola := range p.Escolas {
		if escola == e {
			return
		}
	}
	p.Escolas= append(p.Escolas, e)
	e.AdicionarProfessor(p)
}

func (p Professor) Info() string {
	escolasNomes := ""
	for i, escola := range p.Escolas {
		escolasNomes += escola.Nome
		if i < len(p.Escolas)-1 {
			escolasNomes += ", "
		}
	}
	return fmt.Sprintf("Professor %s leciona nas escolas: %s", p.Nome, escolasNomes)
}

func (e*Escola) AdicionarProfessor(p *Professor) {
	for _, professor := range e.Professores {
		if professor == p {
			return
		}
	}
	e.Professores = append(e.Professores, p)
	p.AdicionarEscola(e)
}

func (e Escola) Info() string {
	professoresNomes := ""
	for i, professor := range e.Professores {
		professoresNomes += professor.Nome
		if i < len(e.Professores)-1 {
			professoresNomes+=", "
		}
	}
	return fmt.Sprintf("Escola %s tem os professores: %s", e.Nome, professoresNomes)
}

func main() {
	// Criar professores
	prof1 := &Professor{Nome: "João"}
	prof2 := &Professor{Nome: "Maria"}

	// Criar escolas
	escola1 := &Escola{Nome: "Escola A"}
	escola2 := &Escola{Nome: "Escola B"}

	// Associação entre escolas e professores
	escola1.AdicionarProfessor(prof1)
	escola2.AdicionarProfessor(prof1)
	prof2.AdicionarEscola(escola1)

	// Exibir informações
	fmt.Println(escola1.Info())
	fmt.Println(escola2.Info())
	fmt.Println(prof1.Info())
	fmt.Println(prof2.Info())
}